README.md
<a href="#table-of-contents">
<img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/documents-with-footer-to-pdf/readme/headlines/documents-with-footer-to-pdf.svg" height="55px" name="Documents with Footer to Pdf for Ruby" alt="# Documents with Footer to .pdf for Ruby">
</a>

Add a footer to each document and create a single .pdf file all in one command.
<br>
<br>
<br>
<a href="#table-of-contents">
<img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/documents-with-footer-to-pdf/readme/headlines/examples.svg" height="55px" name="examples" alt="Examples">
</a>

### Example 1

**Filename**: 
```5--HEADING--Subheadline.png```

```ruby
    { 
        footer__table__left__top__text: '<<--ENUMERATOR_INTEGER-->>. <<--HEADLINE-->>',
        footer__table__left__bottom__text: '<<--TIMESTAMP-->>',
        footer__table__center__top__text: '',
        footer__table__center__bottom__text: '',
        footer__table__right__top__text: '<<--PAGE_CURRENT-->>',
        footer__table__right__bottom__text: '',
    }
```

**Output**

<img src="https://github.com/a6b8/a6b8/blob/main/docs/documents-with-footer-to-pdf/readme/examples/a.jpg?raw=true">


### Example 2

**Filename**: 
```5--HEADING--Subheadline.png```

```ruby
    { 
        footer__table__left__top__text: '<<--ENUMERATOR_ROMAN-->>. <<--HEADLINE-->>',
        footer__table__left__bottom__text: '<<--SUBHEADING-->>',
        footer__table__center__top__text: 'Application',
        footer__table__center__bottom__text: 'John Doe',
        footer__table__right__top__text: '<<--TIMESTAMP-->>',
        footer__table__right__bottom__text: '<<--PAGE_CURRENT-->> of <<--PAGE_TOTAL-->>',
        selectors__timestamp__strf: '%A, %e %B %Y'
    }
```
**Output**

<img src="https://github.com/a6b8/a6b8/blob/main/docs/documents-with-footer-to-pdf/readme/examples/b.jpg?raw=true">


### Example 3

**Filename**: 
```5--HEADING--Subheadline.png```

```ruby
    { 
        footer__table__left__top__text: '<<--TIMESTAMP-->>',
        footer__table__left__bottom__text: '',
        footer__table__center__top__text: '<<--ENUMERATOR_CHAR-->> <<--HEADLINE-->>',
        footer__table__center__bottom__text: '',
        footer__table__right__top__text: '<<--PAGE_CURRENT-->>',
        footer__table__right__bottom__text: '',
    }
```

<img src="https://github.com/a6b8/a6b8/blob/main/docs/documents-with-footer-to-pdf/readme/examples/c.jpg?raw=true">

<br>
<br>
<a href="#table-of-contents">
<img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/documents-with-footer-to-pdf/readme/headlines/table-of-contents.svg" height="55px" name="table-of-contents" alt="Table of Contents">
</a>
<br>

1. [Quickstart](#quickstart)<br>
2. [Parameters](#parameters)<br>
3. [Options](#options)<br>
4. [Selectors](#selectors)<br>
5. [Contributing](#contributing)<br>
6. [Limitations](#limitations)<br>
7. [Credits](#credits)<br>
8. [License](#license)<br>
9. [Code of Conduct](#code-of-conduct)<br>
10. [Support my Work](#support-my-work)<br>

<br>
<br>
<a href="#table-of-contents">
<img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/documents-with-footer-to-pdf/readme/headlines/quickstart.svg" height="55px" name="quickstart" alt="Quickstart">
</a>

```ruby
    require 'documents_with_footer_to_pdf'


    your_folder = './1-test/'

    DocumentsWithFooterToPdf.generate( 
        your_folder, 
        :short, 
        {}
) 
```
<br>
<br>
<a href="#table-of-contents">
<img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/local-path-builder-for-ruby/readme/headlines/setup.svg" height="55px" name="setup" alt="Setup">
</a>

Add this line to your application's Gemfile:

```ruby
gem 'documents_with_footer_to_pdf'
```

And then execute:

    $ bundle install

Or install it yourself as:

    $ gem install documents_with_footer_to_pdf


On Rubygems: 
- Gem: https://rubygems.org/gems/test
- Profile: https://rubygems.org/profiles/a6b8

<br>
<br>
<a href="#table-of-contents">
<img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/documents-with-footer-to-pdf/readme/headlines/methods.svg" height="55px" name="methods" alt="Methods">
</a>

### .get_options()
```ruby
    require 'documents_with_footer_to_pdf'

    hash = DocumentsWithFooterToPdf.get_options() 
    # => { path: 
```


### .generate()
```ruby
    require 'documents_with_footer_to_pdf'

    hash = DocumentsWithFooterToPdf.generate( 
        folder, 
        console_mode,  
        options 
    )
```

**Input**
| **Type** | **Required** | **Description** | **Example** | **Description** |
|------:|:------|:------|:------|:------| 
| **headline** | ```String``` | Yes | "./test/" | Define path to folder |
| **console mode** | ```Symbol``` | Yes | ```:hash``` | Set test console output mode. Use ```:silent```, ```:short```, or ```:detail``` |
| **options** | ```'Hash'``` | No | ```{}``` | Change defaut options, please refer: <a href="#options">Options</a> for more info |

**Return**<br>
Boolean   
<br>
<br>
<br>
<a href="#table-of-contents">
<img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/documents-with-footer-to-pdf/readme/headlines/options.svg" height="55px" name="options" alt="Options">
</a>

### Path
| Nr | Name | Key | Default | Type | Description |
| :-- | :-- | :-- | :-- | :-- | :-- |
| A.1. | Children Tmp Name |:path__children__tmp__name | `"tmp-{{SALT}}"` | String | Change temporary folder name, {{SALT}} will replaced with current unix timestamp |
| A.2. | Children Pdf_combined Name |:path__children__pdf_combined__name | `"0-result-{{SALT}}"` | String | Change result folder name, {{SALT}} will replaced with current unix timestamp |


### Footer
| Nr | Name | Key | Default | Type | Description |
| :-- | :-- | :-- | :-- | :-- | :-- |
| B.1. | Position Top |:footer__position__top | `[0, 20]` | Array | Set footer position top. |
| B.2. | Position Bottom |:footer__position__bottom | `[0, 10]` | Array | Set footer position bottom. |
| B.3. | Table Left Top Text |:footer__table__left__top__text | `"<<--FILENAME-->>"` | String | Set text in field left-top |
| B.4. | Table Left Bottom Text |:footer__table__left__bottom__text | `""` | String | String | Set text in field left-bottom |
| B.5. | Table Center Top Text |:footer__table__center__top__text | `""` | String | String | Set text in field center-top |
| B.6. | Table Center Bottom Text |:footer__table__center__bottom__text | `""` | String | String | Set text in field center-bottom |
| B.7. | Table Right Top Text |:footer__table__right__top__text | `"<<--TIMESTAMP-->>"` | String | String | Set text in field right-top |
| B.8. | Table Right Bottom Text |:footer__table__right__bottom__text | `"<<--PAGE_CURRENT-->> from <<--PAGE_TOTAL-->>"` | String | String | Set text in field right-bottom |


### Selectors
| Nr | Name | Key | Default | Type | Description |
| :-- | :-- | :-- | :-- | :-- | :-- |
| C.1. | Timestamp Gsub |:selectors__timestamp__gsub | `"<<--TIMESTAMP-->>"` | String | Change marker for timestamp. |
| C.2. | Timestamp Strf |:selectors__timestamp__strf | `"%d.%m.%Y"` | String | Change format of date and time |
| C.3. | Page_current Gsub |:selectors__page_current__gsub | `"<<--PAGE_CURRENT-->>"` | String | Change marker for current page. |
| C.4. | Page_total Gsub |:selectors__page_total__gsub | `"<<--PAGE_TOTAL-->>"` | String | Change marker for total page size |
| C.5. | Enumerator_original Gsub |:selectors__enumerator_original__gsub | `"<<--ENUMERATOR_ORIGINAL-->>"` | String | Change marker for original enumeration |
| C.6. | Enumerator_integer Gsub |:selectors__enumerator_integer__gsub | `"<<--ENUMERATOR_INTEGER-->>"` | String | Change marker for integer enumeration |
| C.7. | Enumerator_char Gsub |:selectors__enumerator_char__gsub | `"<<--ENUMERATOR_CHAR-->>"` | String | Change marker for char enumeration |
| C.8. | Enumerator_roman Gsub |:selectors__enumerator_roman__gsub | `"<<--ENUMERATOR_ROMAN-->>"` | String | Change marker for roman letters enumeration |
| C.9. | Filename Gsub |:selectors__filename__gsub | `"<<--FILENAME-->>"` | String | Change marker for filename |
| C.10. | Path Gsub |:selectors__path__gsub | `"<<--PATH-->>"` | String | Change marker for full path |
| C.11. | Heading Gsub |:selectors__heading__gsub | `"<<--HEADLINE-->>"` | String | Change marker for headline |
| C.12. | Subheading Gsub |:selectors__subheading__gsub | `"<<--SUBHEADING-->>"` | String | Change marker for sub headline |

More Information on ```strf``` format: https://apidock.com/ruby/DateTime/strftime

### Params
| Nr | Name | Key | Default | Type | Description |
| :-- | :-- | :-- | :-- | :-- | :-- |
| D.1. | Footer Font_size |:params__footer__font_size | `9` | Integer | Set footer font size |
| D.2. | Document Width |:params__document__width | `500` | Integer | Set document width |
| D.3. | Image Density |:params__image__density | `300` | Integer | Set image density |
| D.4. | Search Subfolders |:params__search__subfolders | `false` | Boolean | Set if subfolders should be included in search |
| D.5. | Search Suffixs |:params__search__suffixs | `["jpg", "png", "pdf"]` | Array | Types of suffixes which are included (Supported jpg, png, pdf) |
| D.6. | Console Length |:params__console__length | `50` | Integer | Change length of console output |
</a>
<br>
<br>
<a href="#table-of-contents">
<img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/documents-with-footer-to-pdf/readme/headlines/selectors.svg" height="55px" name="selectors" alt="Selectors">
</a>

| Selector | Example | Description |
| :--- | :--- | :--- |
| ```"<<--TIMESTAMP-->>"``` | 24.05.2021 | Show current Timestamp, to change for format use option: :selectors__timestamp__strf |
| ```"<<--PAGE_CURRENT-->>"``` | ```1``` | Show current Page |
| ```"<<--PAGE_TOTAL-->>"``` | ```3``` | Show total Page size |
| ```"<<--ENUMERATOR_ORIGINAL-->>"``` | ```"C"``` | Show original Enumerator |
| ```"<<--ENUMERATOR_INTEGER-->>"``` | ```3``` | Show Enumerator as Integer |
| ```"<<--ENUMERATOR_CHAR-->>"``` | ```"C"``` | Show Enumerator as Char |
| ```"<<--ENUMERATOR_ROMAN-->>"``` | ```"III"``` | Show Enumerator in Roman Letters |
| ```"<<--FILENAME-->>"``` | ```"C--HEADING--Subheadline.png"``` | Show filename |
| ```"<<--PATH-->>"``` | ```"../C--HEADING--Subheadline"``` | Show file path|
| ```"<<--HEADLINE-->>"``` | ```"HEADING"``` | Show Headline |
| ```"<<--SUBHEADING-->>"``` | ```"Subheading"``` | Show Subheading |

### Filename Selectors
Use double hyphens '--' as limiter 

**Struct**:
#{SINGLE DIGIT/CHAR}--#{HEADLINE}--#{SUBHEADING}.suffix

**Example Filename**: ```C--HEADING--Subheadline.png```

<br>
<br>
<a href="#table-of-contents">
<img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/documents-with-footer-to-pdf/readme/headlines/contributing.svg" height="55px" name="contributing" alt="Contributing">
</a>

Bug reports and pull requests are welcome on GitHub at https://github.com/a6b8/documents-with-footer-to-pdf. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/a6b8/documents-with-footer-to-pdf/blob/master/CODE_OF_CONDUCT.md).
<br>
<br>
<br>
<a href="#table-of-contents">
<img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/documents-with-footer-to-pdf/readme/headlines/limitations.svg" height="55px" name="limitations" alt="Limitations">
</a>
- Made for local usage
- Some pdf fonts are not supported.
<br>
<br>

<a href="#table-of-contents">
<img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/documents-with-footer-to-pdf/readme/headlines/credits.svg" height="55px" name="credits" alt="Credits">
</a>

This gem depend on following gems:
- **local_path_builder** <br>
  https://github.com/a6b8/local-path-builder-for-ruby <br>
- **FileUtils** <br>
  https://ruby-doc.org/stdlib-2.4.1/libdoc/fileutils/rdoc/FileUtils.html <br>
- **combine_pdf** <br>
  https://github.com/boazsegev/combine_pdf <br>
- **prawn** <br>
  https://github.com/prawnpdf/prawn <br>
- **image magick** <br>
  https://imagemagick.org/script/convert.php
<br>
<br>

<a href="#table-of-contents">
<img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/documents-with-footer-to-pdf/readme/headlines/license.svg" height="55px" name="license" alt="License">
</a>

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).
<br>
<br>
<br>
<a href="#table-of-contents">
<img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/documents-with-footer-to-pdf/readme/headlines/code-of-conduct.svg" height="55px" name="code-of-conduct" alt="Code of Conduct">
</a>
    
Everyone interacting in the documents-with-footer-to-pdf project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/a6b8/documents-with-footer-to-pdf/blob/master/CODE_OF_CONDUCT.md).
<br>
<br>
<br>
<a href="#table-of-contents">
<img href="#table-of-contents" src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/documents-with-footer-to-pdf/readme/headlines/support-my-work.svg" height="55px" name="support-my-work" alt="Support my Work">
</a>
    
Donate by [https://www.paypal.com](https://www.paypal.com/donate?hosted_button_id=XKYLQ9FBGC4RG)

CHANGELOG.md
## [Unreleased]

## [0.1.0] - 2021-06-21

- Initial release


test/t.rb
puts "1.\tFind Docuemnts"
print "\tProcessing\t"

for index in 0..100
    if index%30 == 0 and index != 0
        print( "\n\t\t\t")
    end
    print( '.')
end

test/test.rb
require '../lib/documents_with_footer_to_pdf'

p = './1-test/'

DocumentsWithFooterToPdf.generate( 
    p, 
    :short, 
    { footer__table__right__top__text: 'TEST' }
) 



test/allow.rb
def get_allow_list( params )
    lists = {}
    lists[:allow] = []
    lists[:block] = []

    params.keys.each do | lvl1 |
        if params[ lvl1 ].class.to_s.eql? 'Hash'
            params[ lvl1 ].keys.each do | lvl2 |
                if params[ lvl1 ][ lvl2 ].class.to_s.eql? 'Hash'
                    params[ lvl1 ][ lvl2 ].keys.each do | lvl3 |
                        if params[ lvl1 ][ lvl2 ][ lvl3 ].class.to_s.eql? 'Hash'
                            params[ lvl1 ][ lvl2 ][ lvl3 ].keys.each do | lvl4 |
                                if params[ lvl1 ][ lvl2 ][ lvl3 ][ lvl4 ].class.to_s.eql? 'Hash'
                                    params[ lvl1 ][ lvl2 ][ lvl3 ][ lvl4 ].keys.each do | lvl5 |
                                        if params[ lvl1 ][ lvl2 ][ lvl3 ][ lvl4 ][ lvl5 ].class.to_s.eql? 'Hash'
                                            params[ lvl1 ][ lvl2 ][ lvl3 ][ lvl4 ][ lvl5 ].keys.each do | lvl6 |
                                                if params[ lvl1 ][ lvl2 ][ lvl3 ][ lvl4 ][ lvl5 ].class.to_s.eql? 'Hash'
                                                    puts 'HERE'
                                                else
                                                    str = ( lvl1.to_s + '__' + lvl2.to_s + '__' + lvl3.to_s + '__' + lvl4.to_s ).to_sym
                                                    if !params[ lvl1 ][ lvl2 ][ lvl3 ][ lvl4 ].nil?
                                                        lists[:allow].push( str )
                                                    else
                                                        lists[:block].push( str )
                                                    end
                                                end
                                            end
                                            puts 'HERE'
                                        else
                                            str = ( lvl1.to_s + '__' + lvl2.to_s + '__' + lvl3.to_s + '__' + lvl4.to_s ).to_sym
                                            if !params[ lvl1 ][ lvl2 ][ lvl3 ][ lvl4 ].nil?
                                                lists[:allow].push( str )
                                            else
                                                lists[:block].push( str )
                                            end
                                        end
                                    end
                                    puts 'HERE'
                                else
                                    str = ( lvl1.to_s + '__' + lvl2.to_s + '__' + lvl3.to_s + '__' + lvl4.to_s ).to_sym
                                    if !params[ lvl1 ][ lvl2 ][ lvl3 ][ lvl4 ].nil?
                                        lists[:allow].push( str )
                                    else
                                        lists[:block].push( str )
                                    end
                                end
                            end
                        else
                            str = ( lvl1.to_s + '__' + lvl2.to_s + '__' + lvl3.to_s ).to_sym
                            if !params[ lvl1 ][ lvl2 ][ lvl3 ].nil?
                                lists[:allow].push( str )
                            else
                                lists[:block].push( str )
                            end
                        end
                    end
                else
                    str = ( lvl1.to_s + '__' + lvl2.to_s ).to_sym
                    if !params[ lvl1 ][ lvl2  ].nil? 
                        lists[:allow].push( str )
                    else
                        lists[:block].push( str )
                    end
                end
            end
        else
            str = lvl1
            if !params[ lvl1 ].nil?
                lists[:allow].push( str )
            else
                lists[:block].push( str )
            end
        end
    end
    return lists
end

require '/Users/andreasbanholzer/PROJEKTE/2021-06-documents-with-footer-to-pdf/1/documents_with_footer_to_pdf/lib/documents_with_footer_to_pdf.rb'



lib/documents_with_footer_to_pdf.rb
# frozen_string_literal: true

require_relative "documents_with_footer_to_pdf/version"
require 'local_path_builder'
require 'FileUtils'
require 'combine_pdf'
require 'prawn'
Prawn::Fonts::AFM.hide_m17n_warning = true

module DocumentsWithFooterToPdf
  class Error < StandardError; end
  # Your code goes here...

  def self.get_options()
    return {
      path: {
        root: nil,
        name: '',
        children: {
          tmp: {
            name: 'tmp-{{SALT}}',
            children: {
              jpg: {
                name: '0-jpg'
              },
                pdf_single: {
                  name: '1-pdf-single--w-footer'
              }
            }
          },
          pdf_combined: {
            name: '0-result-{{SALT}}',
            files: {
              result: {
                name: '0-result-{{SALT}}.pdf'
              }
            }
          }
        }
      },
      footer: {
        position: {
          top: [ 0, 20 ],
          bottom: [ 0, 10 ]
        },
        table: {
          left: {
            top: {
              text: '<<--FILENAME-->>'
            },
            bottom: {
              text: ''
            }
          },
          center: {
            top: {
              text: ''
            },
            bottom: {
              text: ''
            }
          },
          right: {
            top: {
              text: '<<--TIMESTAMP-->>'
            },
            bottom: {
              text: '<<--PAGE_CURRENT-->> from <<--PAGE_TOTAL-->>'
            }
          }
        }
      },
      selectors: {
        timestamp: {
          gsub: '<<--TIMESTAMP-->>',
          key: :timestamp,
          strf: '%d.%m.%Y',
        },
        page_current: {
          gsub: '<<--PAGE_CURRENT-->>',
          key: :page__current
        },
        page_total: {
          gsub: '<<--PAGE_TOTAL-->>',
          key: :page__total
        },
        enumerator_original: {
          gsub: '<<--ENUMERATOR_ORIGINAL-->>',
          key: :enumerator__original
        },
        enumerator_integer: {
          gsub: '<<--ENUMERATOR_INTEGER-->>',
          key: :enumerator__integer
        },
        enumerator_char: {
          gsub: '<<--ENUMERATOR_CHAR-->>',
          key: :enumerator__char
        },
        enumerator_roman: {
          gsub: '<<--ENUMERATOR_ROMAN-->>',
          key: :enumerator__roman
        },
        filename: {
          gsub: '<<--FILENAME-->>',
          key: :filename
        },
        path: {
          gsub: '<<--PATH-->>',
          key: :path
        },
        heading: {
          gsub: '<<--HEADLINE-->>',
          key: :heading
        },
        subheading: {
          gsub: '<<--SUBHEADING-->>',
          key: :subheading
        }
      },
      params: {
        footer: {
          font_size: 9
        },
        document: {
          width: 500
        },
        image: {
          density: 300
        },
        search: {
          subfolders: false,
          suffixs: [ 'jpg', 'png', 'pdf' ]
        },
        console: {
          silent: nil,
          mode: nil,
          length: 50
        }
      }
    }
  end


  def self.generate( folder, silent, options={} )
    
    hash = Marshal.load( Marshal.dump( self.get_options() ) )
    if self.validate_generate( folder, silent, options, hash )
            
      hash[:path][:root] = folder
      hash[:params][:console][:silent] = silent == :silent ? true : false
      hash[:params][:console][:mode] = !hash[:params][:console][:silent] ? silent : ''
      hash[:path] = LocalPathBuilder.generate( hash[:path], :silent, Time.now.to_i.to_s )
      hash = self.options_update( options, hash, 'set_options' )  

      self.footer_image( hash )
      prepares = self.footer_prepare( hash )
      self.footer_generate( prepares, hash )
      self.footer_merge( hash )
      FileUtils.rm_rf( hash[:path][:children][:tmp][:full] )
    end
  end


  private


  def self.debug( obj )
    return obj[:params][:console][:silent]
  end


  def self.options_update( vars, template, mode )
    def self.str_difference( a, b )
      a = a.to_s.downcase.split( '_' ).join( '' )
      b = b.to_s.downcase.split( '_' ).join( '' )
      longer = [ a.size, b.size ].max
      same = a
        .each_char
        .zip( b.each_char )
        .select { | a, b | a == b }
        .size
      ( longer - same ) / a.size.to_f
    end


    allow_list = [
      :path__children__tmp__name,
      :path__children__pdf_combined__name,
      :footer__position__top,
      :footer__position__bottom,
      :footer__table__left__top__text,
      :footer__table__left__bottom__text,
      :footer__table__center__top__text,
      :footer__table__center__bottom__text,
      :footer__table__right__top__text,
      :footer__table__right__bottom__text,
      :selectors__timestamp__gsub,
      :selectors__timestamp__strf,
      :selectors__page_current__gsub,
      :selectors__page_total__gsub,
      :selectors__enumerator_original__gsub,
      :selectors__enumerator_integer__gsub,
      :selectors__enumerator_char__gsub,
      :selectors__enumerator_roman__gsub,
      :selectors__filename__gsub,
      :selectors__path__gsub,
      :selectors__heading__gsub,
      :selectors__subheading__gsub,
      :params__footer__font_size,
      :params__document__width,
      :params__image__density,
      :params__search__subfolders,
      :params__search__suffixs,
      :params__console__length
    ]

    messages = []
    _options = Marshal.load( Marshal.dump( template ) )
    
    vars.keys.each do | key |
      if allow_list.include?( key ) 
  
        keys = key.to_s.split( '__' ).map { | a | a.to_sym }
        case( keys.length )
          when 1
            _options[ keys[ 0 ] ] = vars[ key ]
          when 2
            _options[ keys[ 0 ] ][ keys[ 1 ] ] = vars[ key ]
          when 3
            _options[ keys[ 0 ] ][ keys[ 1 ] ][ keys[ 2 ] ] = vars[ key ]
          when 4
            _options[ keys[ 0 ] ][ keys[ 1 ] ][ keys[ 2 ] ][ keys[ 3 ] ] = vars[ key ]
          when 5
            _options[ keys[ 0 ] ][ keys[ 1 ] ][ keys[ 2 ] ][ keys[ 3 ] ][ keys[ 4 ] ] = vars[ key ]
          when 6
            _options[ keys[ 0 ] ][ keys[ 1 ] ][ keys[ 2 ] ][ keys[ 3 ] ][ keys[ 4 ] ][ keys[ 5   ] ] = vars[ key ]
        end
      else
        nearest = allow_list
          .map { | word | { score: self.str_difference( key, word ), word: word } }
          .min_by { | item | item[:score] }

        message = "\"#{key}\" is not a valid key, did you mean \"<--similar-->\"?"
        message = message.gsub( '<--similar-->', nearest[:word].to_s )
        messages.push( message )
      end
    end
    
    result = nil
    case mode
      when 'check_options'
        result = messages
      when 'set_options'
        result = _options
    end

    return result
  end


  def self.validate_generate( folder, silent, vars, template )
    messages = {
      folder: [],
      silent: [],
      options: [],
      other: []
   }

    if folder.class.to_s.eql?( 'String' )
      if File.directory?( folder )

      else
        messages[:folder].push( 'Is not a valid path or not exist.')
      end
    else
      messages[:folder].push( 'Is not Type "String"')
    end

    if silent.class.to_s.eql?( 'Symbol' )
      if [:silent, :short, :detail].include?( silent )

      else
        messages[:silent].push( 'Is not :silent, :short or :detail')
      end
    else
      messages[:silent].push( 'Is not Type "Symbol"')
    end

    if vars.class.to_s.eql?( 'Hash' )
      messages[:options] = self.options_update( vars, template, 'check_options' )
    else
      messages[:options].push( 'Is not Type "Hash".') 
    end

    valid = messages.keys.map { | key | messages[ key ].length }.sum == 0

    if !valid
      puts 'Following errors occured:'
      messages.keys.each do | key |
        if messages[ key ].length != 0
          puts "  #{key[ 0, 1 ].upcase}#{key[ 1, key.length ]}"
          messages[ key ].each do | m |
            puts "  - #{m}"
          end
        end
      end
    end
    return valid
  end


  def self.footer_image( obj )
    files = []

    star = '*'
    obj[:params][:search][:subfolders] ? star = '**/*' : ''
    obj[:params][:search][:suffixs]
      .each { | a | files += Dir[ "#{obj[:path][:full]}#{star}.#{a}" ].sort }

    t = files
      .map { | a | a.split( '.' ).last }
      .inject( Hash.new( 0 ) ) { |total, e| total[e] += 1 ;total }
      .map { | a | ".#{a[ 0 ]} #{a[ 1 ]}" }
      .join(', ')

    !self.debug( obj ) ? print( "1. \tFind Files (#{files.length} Total | #{t}):" ) : ''

    files.each.with_index do | file, index |
        a = obj[:params][:image][:density].to_s
        b = obj[:path][:children][:tmp][:children][:jpg][:full]
        c = File.basename( file ).split( '.' )[ 0..-2 ].join( '.' )
        cmd = "convert -density #{a} -background white -alpha remove \"#{file}\" \"#{b}#{c}.jpg\""

        out = IO.popen( cmd )
        out.readlines

        self.console_mode( file, index, obj )
    end

    !self.debug( obj ) ? puts : ''

    return true
  end


  def self.footer_prepare( obj )
    def self.roman_numerals( integer )
      def self.next_lower_key( integer, values )
        arabics = values.keys
        next_lower_index = ( arabics.push( integer ).sort.index( integer ) ) - 1
        arabics[next_lower_index]
      end
      
      values = {
        1 => 'I',
        4 => 'IV',
        5 => 'V',
        9 => 'IX',
        10 => 'X',
        40 => 'XL',
        50 => 'L',
        90 => 'XC',
        100 => 'C',
        400 => 'CD',
        500 => 'D',
        900 => 'CM',
        1000 => 'M'
      }
      
      roman = ''
      while integer > 0
        if values[ integer ]
          roman += values[ integer ]
          return roman
        end
    
        roman += values[ self.next_lower_key( integer, values ) ]
        integer -= self.next_lower_key( integer, values )
      end
    end
    
    
    def self.create_selectors( str, item, alpha )
      if str.index( '--' ) == 1
        item[:valid] = true
        a = str.split( '--' )
        item[:enumerator][:original] = a[ 0 ]
        case a.length
          when 2
            item[:heading] = a[ 1 ]
          when 3
            item[:heading] = a[ 1 ]
            item[:subheading] = a[ 2 ]
        end
      
        if /\A[-+]?[0-9]+\z/.match( item[:enumerator][:original] ).to_a.length != 0
          item[:enumerator][:style] = :integer
          item[:enumerator][:integer] = item[:enumerator][:original].to_i
          item[:enumerator][:char] = alpha[ item[:enumerator][:integer]-1 ]
          item[:enumerator][:roman] = self.roman_numerals( item[:enumerator][:integer] )
        else
          item[:enumerator][:style] = :char
          item[:enumerator][:char] = item[:enumerator][:original].upcase
          item[:enumerator][:integer] =alpha.index( item[:enumerator][:char] ) + 1
          item[:enumerator][:roman] = self.roman_numerals( alpha.index( item[:enumerator][:char] ) + 1 )
        end
      else
        item[:valid] = false
        item[:heading] = str
      end
      
      return item
    end

    timestamp = Time.now.strftime( obj[:selectors][:timestamp][:strf] )
    paths = Dir[ obj[:path][:children][:tmp][:children][:jpg][:full] + '*' ]
    pages = []

    alpha = ( 'A'.upto( 'Z' ) ).to_a
    for i in 0..paths.length - 1
      path = paths[ i ]
      
      p = {
        file: {
          from: nil,
          to: nil
        },
        selectors: {
          valid: nil,
          timestamp: nil,
          heading: nil,
          filename: nil,
          path: nil,
          subheading: nil,
          enumerator: {
            style: nil,
            original: nil,
            char: nil,
            integer: nil,
            roman: nil
          },
          page: {
            current: nil,
            total: nil
          }
        }
      }
  
      p[:file][:from] = path

      name = File.basename( path ).split( '.' )[ 0..-2 ].join( '.' )
      tmp = obj[:path][:children][:tmp][:children][:pdf_single][:full]
      p[:file][:to] = "#{tmp}#{name}.pdf"

      p[:selectors] = self.create_selectors( name, p[:selectors], alpha )
      p[:selectors][:timestamp] = timestamp 
      p[:selectors][:filename] = path.split('/').last
      p[:selectors][:path] = path
      p[:selectors][:page][:current] = ( i + 1 ).to_s
      p[:selectors][:page][:total] = paths.length.to_s
      pages.push( p )
    end

    return pages
  end


  def self.footer_generate( prepares, obj )
    !self.debug( obj ) ? print( "2.\tWrite Footer:" ) : ''

    prepares.each.with_index do | prepare, index |
      footer = Marshal.load( Marshal.dump( obj[:footer] ) )
      footer[:table].keys.each do | align |
        footer[:table][ align ].keys.each do | position |
          sels = obj[:selectors].keys.map { | a | [ obj[:selectors][ a ][:gsub], obj[:selectors][ a ][:key] ] }
          sels.each do | sel |
            if footer[:table][ align ][ position ][:text].include? sel[ 0 ]
              keys = sel[ 1 ].to_s.split( '__' ).map { | a | a.to_sym }
              case keys.length
                when 1
                  footer[:table][ align ][ position ][:text] = footer[:table][ align ][ position ][:text]
                    .gsub( sel[ 0 ], prepare[:selectors][ keys[ 0 ] ].to_s )
                when 2
                  footer[:table][ align ][ position ][:text] = footer[:table][ align ][ position ][:text]
                    .gsub( sel[ 0 ], prepare[:selectors][ keys[ 0 ] ][ keys[ 1 ] ].to_s )
              end
            end
          end
        end
      end

      Prawn::Document.generate( prepare[:file][:to] ) do | pdf |
        pdf.image(
          prepare[:file][:from],
          at: [ 0, pdf.bounds.top ],
          width: obj[:params][:document][:width]
        )

        footer[:table].keys.each do | align |
          footer[:table][ align ].keys.each do | position |
            pdf.text_box(
              footer[:table][ align ][ position ][:text],
              size: obj[:params][:footer][:font_size],
              align: align,
              at: footer[:position][ position ],
            )
          end
        end
      end

      self.console_mode( prepare[:file][:from], index, obj )
    end

    !self.debug( obj ) ? puts : ''

    return true
  end


  def self.footer_merge( obj )
    !self.debug( obj ) ? puts( "3.\tMerge" ) : ''
    
    files = Dir[ obj[:path][:children][:tmp][:children][:pdf_single][:full] + '*.pdf' ].sort
    
    pdf = CombinePDF.new
    files.each do | file |
      pdf << CombinePDF.load( file )
    end

    p = obj[:path][:children][:pdf_combined][:files][:result][:full]
    pdf.save( p )
    
    return true
  end


  def self.console_mode( file, index, obj )
    if !self.debug( obj )
      case obj[:params][:console][:mode]
        when :short
          index%obj[:params][:console][:length] == 0 ? print( "\n\t" ) : ''
          print( '.' )
        when :detail
          puts ( "\t" )
          p = ''
          if file.length > obj[:params][:console][:length]
            _start = obj[:params][:console][:length] / 5
            _end = obj[:params][:console][:length] - _start
            p += file[ 0, _start ]
            p += '...'
            p += file[ file.length - _end, file.length]
          else
            p = file
          end
          print( "\t- #{p}" )
      end
    else
    end
  end
end
